detectable/src/main/java/com/synopsys/integration/detectable/detectables/bazel/pipeline/step/FinalStepJsonProtoHaskellCabalLibraries.java

```diff
@@ -0,0 +1,146 @@
+
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.synopsys.integration.bdio.graph.MutableDependencyGraph;
+import com.synopsys.integration.bdio.graph.MutableMapDependencyGraph;
+import com.synopsys.integration.bdio.model.Forge;
+import com.synopsys.integration.bdio.model.dependency.Dependency;
+import com.synopsys.integration.bdio.model.externalid.ExternalId;
+import com.synopsys.integration.bdio.model.externalid.ExternalIdFactory;
+import com.synopsys.integration.exception.IntegrationException;
+import com.synopsys.integration.util.NameVersion;
+
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+
+public class FinalStepJsonProtoHaskellCabalLibraries implements FinalStep {
+    private static final String FORGE_NAME = "hackage";
+    private static final String FORGE_SEPARATOR = "/";
+    private final Logger logger = LoggerFactory.getLogger(this.getClass());
+    private final Forge hackageForge = new Forge(FORGE_SEPARATOR, FORGE_NAME);
+
+    @Override
+    public MutableDependencyGraph finish(final List<String> input) throws IntegrationException {
+        final String jsonString = extractJsonString(input);
+        final JsonArray resultsArray = parseResultsJson(jsonString);
+        final MutableDependencyGraph dependencyGraph = new MutableMapDependencyGraph();
+        for (final JsonElement resultsArrayMemberElement : resultsArray) {
+            final Optional<JsonObject> ruleObject = extractHaskellCabalLibraryRule(resultsArrayMemberElement);
+            if (ruleObject.isPresent()) {
+                addDependencyToGraph(dependencyGraph, ruleObject.get());
+            }
+        }
+        return dependencyGraph;
+    }
+
+    private String extractJsonString(final List<String> input) throws IntegrationException {
+        if (input.size() != 1) {
+            throw new IntegrationException(String.format("Input size is %d; expected 1", input.size()));
+        }
+        return input.get(0);
+    }
+
+    private JsonArray parseResultsJson(final String jsonString) {
+        final JsonElement protoElement = JsonParser.parseString(jsonString);
+        final JsonObject protoObject = protoElement.getAsJsonObject();
+        final JsonElement resultsElement = protoObject.get("results");
+        return resultsElement.getAsJsonArray();
+    }
+
+    private Dependency haskageCompNameVersionToDependency(final String compName, final String compVersion) {
+        final ExternalId externalId = (new ExternalIdFactory()).createNameVersionExternalId(hackageForge, compName, compVersion);
+        externalId.createBdioId(); // Validity check; throws IllegalStateException if invalid
+        return new Dependency(compName, compVersion, externalId);
+    }
+
+    private Optional<JsonObject> extractHaskellCabalLibraryRule(final JsonElement resultsArrayMemberElement) {
+        final JsonObject resultsArrayMemberObject = resultsArrayMemberElement.getAsJsonObject();
```

Comment:
Did you consider creating gson/pojo model classes?

Reply:
I did... Bazel is very unstable (3 major versions in the last 8 months, and they include breaking changes even in minor versions), and I thought this might be more resilient to change. But I'm not actually sure of that... do you happen to know?

Reply:
It probably depends on the changes, any idea what they typically change and will you need to support multiple models at the same time? I'd be hesitant to do the POJO if the changes were deeper than field renames - like if they were collections and the types stored in the collection changes. 

If it is just field name changes, having muliple fields on the POJO some of which might be empty seems reasonable to me, and your code is just defensive against missing fields which it needs to be in the JsonElement case anyway. 

I think

If pojo.myField != null then haskellName = pojo.myField else haskellName = pojo.otherField

seems better than 

if json.containsField("myField") then haskellName = pojo.getFieldAsString("myField") else haskellName = pojo.getFieldAsString("otherField")

I would definitely understand the json element if the two models are dramatically different and cannot be mashed together.

Reply:
changed it to a gson/pojo approach.